### ISSUE TITLE
[Streams Core] Performance Concerns - Is there a roadmap for performance improvements?

### ISSUE THREAD

**Issue by mattcdavis1** [pyrocms/pyrocms#4273]
Currently, Pyro is nearly 3x slower than Craft v3: https://goo.gl/eBL7tZ

The screenshot above is without opcache but the difference is the same either way (they both gain with opcache).  Also, opcache is is not always going to be available for a given page depending on how many sites are on the server, how much memory is available etc.

Here's a comparison with opcache: https://goo.gl/6eegKn . The relative difference stays about the same.

Both comparisons are with xdebug not loaded and both CMSs are running on modern object-oriented frameworks (Yii2 vs Laravel).  The feature set is also comparable with Craft CMS coming out ahead in more areas than not.

Do you foresee significant (2x) performance gains in the near future (6-12 months) or only minor gains (or possibly even increases)?  Is performance a primary consideration in ongoing development and architecture?  This (performance) is currently my biggest concern with Pyro - it is noticeably sluggish compared with Craft CMS.

**Comment by RCrowt** [comment]
I am also having serious performance issues with a clean install on PyroCMS. Page-load times are around 10 seconds but increase to around 25 seconds if I change and cached assets. 

To try diagnose the problem I enabled xDebugs Profiler which squarely puts the performance issues on the Composer Autoloader. I have many websites running on my machine which are dependent on Composer's autoloader and they have no performance issues. Could PyroCMS be doing something out of the ordinary when loading classes to cause these issues?
![profile](https://cloud.githubusercontent.com/assets/10333050/25066434/368b2db6-221d-11e7-8a92-252b9f1c1ba1.PNG)

**Comment by mattcdavis1** [comment]
@RCrowt - if you page load is taking that long there's something else going on here.  First thing is to make sure xdebug is not being loaded (disabling is not enough).  Next, make sure you have `LIVE_ASSETS=false` in your .env file

**Comment by RCrowt** [comment]
@mattcdavis1 - XDebug is fully disabled by which i mean it's commented out in the php.ini. LIVE_ASSETS is false.

My machine has multiple versions of Laravel, WordPress and various other Frameworks installed which execute with no issues. I have 3 installations of PyroCMS and they are all this slow.

I am running the latest version of XAMPP on a quad core Windows 10 machine with an SSD so performance shouldn't be an issue.

**Comment by mattcdavis1** [comment]
What is the page load time for an empty page?  What type of stuff do you have going on on the page your loading?  Are you generating alot of image/file urls?  Can you post the code that is executing (twig and controller)

**Comment by RCrowt** [comment]
I have rebooted my computer a few times and tweaked a few things in XAMPP and things seem to be a lot faster today. Maybe it is something in my config but it's just strange how it only affects PyroCMS installs.

The theme I am developing has 3 small SASS files which contain only 2 image assets. A page reload now take between 2.5 and 5 seconds both running my theme and the default "Starter" theme. I haven't changed the default content yet or installed/developed any other modules.

**Comment by mattcdavis1** [comment]
hmm-  that still sounds too slow. I'm able to get 150-200ms with opcache enabled and 350-400ms without opcache (with the starter theme and the posts loop disabled since i've removed that module).  Can you just make extra sure that xdebug is not loading?  If you print a `phpinfo(); die();` in public/index.php and do a `cmd + f` for the word "xdebug" it should not show up anywhere.  Can you confirm this?

Also, can you try opening the starter theme `layouts/default.twig` file, deleting all it's contents, and then see what your page load time is?

I'm assuming you're not in a VM since you're using XAMPP.  If xdebug is truly disabled i would try running on native mac outside of XAMPP as a next troubleshooting step to see if XAMPP is causing some issues.

**Comment by RyanThompson** [comment]
It's definitely something I'm both aware of and anxious to fix asap. It's kinda one of those things looming about that has to be addressed. I know where most of it comes from as well and it's the autoloader / lookups for the most part. `composer dump-autoload -o` Will help a bit more as it cuts lookups in half.

A crap ton can be done in 12 months and the same for 6 so I'm confident in saying it'll be addressed not only constantly as we go forward but those milestones should display very notable changes.

Part of it too is getting more people involved at that level who can get eyes on it / benchmark with blackfire for example too.

**Comment by RyanThompson** [comment]
@RCrowt are you using SSD? What's you environment? The lookups are disk heavy so vagrant setups with shared disk resources can cause issues with abnormally high load times.

**Comment by Piterden** [comment]
Tiil all time, I using Streams, about half of year, there were many kinds of issues, but performance issues never
![image](https://cloud.githubusercontent.com/assets/5930429/25072175/eda9a190-22d0-11e7-9508-eea9ca9959ad.png)

**Comment by mattcdavis1** [comment]
@RyanThompson - great to hear that performance is a priority moving forward and that you expect significant improvements over the next 6-12 months.  Would be great to see some comparative bench marks at major point releases (3.1, 3.2, 3.3 etc) vs the other major frameworks and the previous Pyro release.  

One good metric, for example, would be the average page load time over 100 requests for a home page with just some simple text on Wordpress, Craft V3 and Drupal.  Something like at the bottom of [this page](https://craftcms.com/3) (except they are only comparing against themselves).  

[Screenshot ](https://goo.gl/r9Tm0g)

BTW - `composer dumpautoload -o` shows no significant performance gains for me.  I could see that helping in VMs though, since filesystem checks are effectively network calls.

**Comment by RCrowt** [comment]
@mattcdavis1 - I have checked `phpinfo() `and xdebug is not in there.

The default theme currently loads in about 2.9 seconds. When I empty the `layouts/default.twig` template the load time gets slightly better at 2.3 seconds.  Just to clarify this is on a clean install (again) and the page load times are as reported by Chrome Developer tools. 

For comparison a native laravel app I have running on my machine loads it's homepage in a just under 1 second. I have a horrible WordPress website which also loads it's homepage in a similar time.

I am running the latest version of XAMPP on a quad core Windows 10 machine with 4gb ram. The PyroCMS, Xampp and Windows are all running from an SSD. 

I am considering moving to Vagrant but haven't had the time to look into it properly yet. I did try `composer dumpautoload -o` but it gave me no significant performance increase.

**Comment by mattcdavis1** [comment]
i think the remaining performance issues are XAMPP issues.  I'm pretty sure you're running into some virtualization slowness which takes a particularly big hit on disk operations. 

You might be able to run on the Windows Linux Subsystem or IIS and get improved performance or possibly there may be other faster virtualization options but i think the main problem now is at a virtualization level rather than application level.  You could confirm this by running the same code on a linux machine.  You should be able to reproduce the same results as i am getting on my mac.

**Comment by RyanThompson** [comment]
@RCrowt I am glad you've found the same issues as I have with the profiling - goes to show I am on the same page. I think my primary concern will/should be (going forward) is optimizing the structure so that the loader's weight is reduced.. I think it's the size of the classmap in there perhaps. Either way I am a little encouraged that your findings are the same as mine.. :-/

**Comment by RCrowt** [comment]
@mattcdavis1 Maybe there are issues with my XAMPP but if there are it should be easy to replicate as it's a standard install. Maybe the XAMPP is amplifying performance issues with PyroCMS which could give some insight when debugging.

@RyanThompson I would be happy to send you my profiler results if it's of any help.

**Comment by Pixney-William** [comment]
@mattcdavis1 tested it out recently and compared to craft? Just asking because it seems like i still got around 300ms. But i might not have the same setup for testing as you.

And speed is very important to me so i am interested to see if there has been improvements made over the last 6 months.

**Comment by Pixney-William** [comment]
I have also tried `composer dumpautoload -o` with no improvements.

Do we have a roadmap or something regarding these kind of improvements @RyanThompson ?

**Comment by zvineyard** [comment]
It wasn't until I upgraded to PyroCMS 3.4 that I really saw the performance on my site decline. At the moment, I'm limited to non-SSD hosting. I've tried enabling opcache, fiddling with NFS, disabling xdebug, etc. I still get about a 5 second delay. Do I just need more processing power? Any other recommendations I should pursue?

**Comment by RCrowt** [comment]
I don't know if this is of any use but I did some performance experimentation last week. I am building a data importer for a project so I benchmarked the various methods of importing data and the results were very interesting..

For the test I inserted 10,000 records from a CSV file and timed how long it took to to iterate and insert the rows...

- **DB:Insert()**  - A simple array passed straight to the DB insurance = 12 Seconds.
- **Eloquent Model** - Create a new model instance and set the properties one at a time in a loop. = 16 Seconds
- **Stream Model** - Create a new streams model instance and set the properties one at a time in a loop (All field types were text). - 8 Minutes, 45 seconds.

Performance may not be a priority for for PyroCMS but it's something I am personally working on. Given Pyros huge dependency on streams and the benchmark above I think it's worth a look in this area. 

I plan to step through the above code in the coming weeks in the hope of finding some bottlenecks.
